### PR TITLE
Refactor evaluations and report

### PR DESCRIPTION
**Pull request by dzeber**
- Redesign the Dataset API in `dataset.py`
    * no longer encapsulates a train/test split of the data
    * original version moved to `deprecated/train_test_dataset.py`
- Drop `base_evaluation.py` for now until design becomes more concrete
    * rename to `_base_evaluation.py`
- Update ClassificationModel in `model.py` to use new Dataset API
- Add `evaluations/conditional_metric.py`
    * refactoring `evaluations/_example_misclass_rate.py` to use more general
      computation from `misclass_rate.py`
    * refactoring of the computation code itself
    * mostly the same plotting code carried over to `display_result()`
    * move `misclassifcations/` to `deprecated/`
- Add `evaluations/conditional_distribution.py` corresponding to
  `conditional_feature_distribution.py`
    * move that to `deprecated/`
    * refactoring of the computation code itself
    * mostly the same plotting code carried over to `display_result()`
- Updates to `report/`:
    * refactor `landing.ipynb` & `setup_report.py` to use new API
    * combine `misclass_rate.ipynb` & `conditional_metrics.ipynb` into `conditional_metric.ipynb`
    * refactor `cond_feature_distr.ipynb` & rename to `conditional_distribution.ipynb`
- Tests for modules in `presc/deprecated/` moved to `tests/deprecated/`
- Add tests for `dataset.py` & `model.py`
- Add synthetic dataset to use for testing
    * code to generate it included in `generate_data.py`

### CODE REVIEW

File: presc/evaluations/conditional_distribution.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+            plt.show(block=False)
+
+
+class ConditionalDistribution:
+    """Computation of data distributions conditional on prediction results.
+
+    model: the ClassificationModel to run the evaluation for
+    test_dataset: a Dataset to use for evaluation.
+    config: optional dict of config options to override the defaults
```
Comment (gleonard-m) [inline presc/evaluations/conditional_distribution.py:150]: Add the defaults here and the possible dict keys (if they are fixed)

File: presc/evaluations/conditional_distribution.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            data_col=self._test_dataset.df[colname],
+            true_labs=self._test_dataset.labels,
+            pred_labs=self._test_pred,
+            as_categorical=as_categorical,
+            config=self._config,
+        )
+
+    def display(self, colnames=None):
+        """Displays the conditional distribution result for each specified column.
+
```
Comment (gleonard-m) [inline presc/evaluations/conditional_distribution.py:181]: Indicate that it computes and displays so there is not confusion about another step involved

**Review by gleonard-m — COMMENTED** [review]
No large changes needed, code is clean and clear.  Minor comments about variable names and documentation.  Plus these items.
- Rename Spatial Distribution.ipynb to follow standard file naming structure
- Update environment.yml to included jupyter-book==0.9.1

File: tests/test_model.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+from sklearn.preprocessing import StandardScaler, OneHotEncoder
+from sklearn.compose import ColumnTransformer, make_column_selector
+from sklearn.exceptions import NotFittedError
+from numpy import arange
+from pandas import Series, DataFrame
+
+from presc.model import ClassificationModel
+from presc.dataset import Dataset
+
+TEST_SET_SELECTOR = arange(100) >= 80
```
[the remark this answers is not in the document]
Comment (dzeber) [reply tests/test_model.py:13]: Oh, no, that's left over from a previous iteration

File: presc/evaluations/conditional_metric.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+        colnames: a list of column names to run the evaluation over, creating a plot
+            for each. If not supplied, defaults to all feature columns.
+        """
+        if colnames is None:
+            colnames = self._test_dataset.feature_names
+        for colname in colnames:
+            # TODO: don't hardcode the metric
+            eval_result = self.compute_for_column(colname, metric=METRIC)
+            eval_result.display_result(xlab=colname, ylab="Metric value")
```
[the remark this answers is not in the document]
Comment (dzeber) [reply presc/evaluations/conditional_metric.py:164]: I passed this in as a function arg. However this will prob get updated further as I've temporarily fixed the metric but that will get passed in as well.

File: presc/evaluations/conditional_metric.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+            true_labs=self._test_dataset.labels,
+            pred_labs=self._test_pred,
+            metric=metric,
+            as_categorical=as_categorical,
+            config=self._config,
+        )
+
+    def display(self, colnames=None):
+        """Displays the conditional metric result for each specified column.
+
```
[the remark this answers is not in the document]
Comment (dzeber) [reply presc/evaluations/conditional_metric.py:155]: Good point :)

File: presc/model.py [hunk 6]
```diff
@@ -1,63 +1,52 @@
...
-        return self._classifier.predict(X_test)
+        pred = self._classifier.predict(test_dataset.features)
+        return Series(pred, index=test_dataset.features.index)
 
-    @property
-    def test_predictions(self):
-        """Returns the predicted labels for the test set."""
-        return self._test_preds
+    def predict_probs(self, test_dataset):
+        """Compute predicted probabilities for the given test dataset.
```
[the remark this answers is not in the document]
Comment (dzeber) [reply presc/model.py:42]: I added a custom exception class and now raise this over the scikit-learn AttributeError so that the error message is more clear. Happy to handle it differently if you have any strong feelings. Also I don't think this function is currently used in the evaluations, so we could rethink this later if necessary.

**Comment by dzeber** [comment]
I've made the updates - feel free to merge if there's nothing else outstanding.

**Review by gleonard-m — APPROVED** [review]
